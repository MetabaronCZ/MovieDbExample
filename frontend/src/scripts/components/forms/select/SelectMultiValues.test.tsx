import { act } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';
import {
  fireEvent,
  getByRole,
  getAllByRole,
  queryByRole,
  getAllByTitle,
} from '@testing-library/dom';

import { TestComponentWrapper, waitForComponent } from 'test-utils/component';
import { SelectMultiValues } from 'components/forms/select/SelectMultiValues';

describe('components/forms/select/SelectMultiValues', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <SelectMultiValues
        id="ID"
        items={[
          { title: 'Item 1', value: 1, description: 'DESC 1' },
          { title: 'Item 2', value: 2, description: 'DESC 2' },
        ]}
        onSearch={() => null}
        onRemove={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(3); // 2 value items + search input list item

    // item 1
    expect(listItems[0]).toHaveTextContent('Item 1');

    const item1Buttons = listItems[0].querySelectorAll('button');
    expect(item1Buttons.length).toEqual(2);
    expect(item1Buttons[0]).toHaveAttribute('title', 'DESC 1');
    expect(item1Buttons[1]).toHaveAttribute('title', 'Odebrat');

    // item 2
    expect(listItems[1]).toHaveTextContent('Item 2');

    const item2Buttons = listItems[1].querySelectorAll('button');
    expect(item2Buttons.length).toEqual(2);
    expect(item2Buttons[0]).toHaveAttribute('title', 'DESC 2');
    expect(item2Buttons[1]).toHaveAttribute('title', 'Odebrat');

    // search input
    const search = getByRole(listItems[2], 'textbox');
    expect(search).toBeInTheDocument();
    expect(search).toHaveAttribute('id', 'ID');
    expect(search).toHaveAttribute('placeholder', 'Hledat');
    expect(search).toHaveAttribute('value', '');
  });

  it('should render wrapped variant', async () => {
    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        wrapped
        onRemove={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
  });

  it('should not render search input when onSearch not defined', async () => {
    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onRemove={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(2);

    const search = queryByRole(container, 'textbox');
    expect(search).not.toBeInTheDocument();
  });

  it('should not render at all when no options and onSearch not defined', async () => {
    const { container } = await waitForComponent(
      <SelectMultiValues items={[]} onRemove={() => null} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
    expect(container.innerHTML).toEqual('');
  });

  it('should remove items', async () => {
    const onRemove = jest.fn();

    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onRemove={onRemove}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onRemove).not.toHaveBeenCalled();

    const removeButtons = getAllByTitle(container, 'Odebrat');
    expect(removeButtons.length).toEqual(2);

    // remove item 2
    act(() => {
      fireEvent.click(removeButtons[1]);
    });

    expect(onRemove).toHaveBeenCalledTimes(1);
    expect(onRemove).toHaveBeenCalledWith(2);
  });

  it('should not remove items when disabled', async () => {
    const onRemove = jest.fn();

    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        disabled
        onRemove={onRemove}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onRemove).not.toHaveBeenCalled();

    const removeButtons = getAllByTitle(container, 'Odebrat');
    expect(removeButtons.length).toEqual(2);

    // remove item 2
    act(() => {
      fireEvent.click(removeButtons[1]);
    });
    expect(onRemove).not.toHaveBeenCalled();
  });

  it('should not remove disabled item', async () => {
    const onRemove = jest.fn();

    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2, disabled: true },
        ]}
        onRemove={onRemove}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onRemove).not.toHaveBeenCalled();

    const removeButtons = getAllByTitle(container, 'Odebrat');
    expect(removeButtons.length).toEqual(2);

    // remove item 2
    act(() => {
      fireEvent.click(removeButtons[1]);
    });
    expect(onRemove).not.toHaveBeenCalled();
  });

  it('should search items', async () => {
    jest.useFakeTimers();

    const onSearch = jest.fn();

    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSearch={onSearch}
        onRemove={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSearch).not.toHaveBeenCalled();

    const search = getByRole(container, 'textbox');
    expect(search).toBeInTheDocument();

    // search query string
    act(() => {
      fireEvent.change(search, { target: { value: 'QUERY' } });
      jest.runAllTimers();
    });

    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith('QUERY');

    jest.useRealTimers();
  });

  it('should not search when disabled', async () => {
    jest.useFakeTimers();

    const onSearch = jest.fn();

    const { container } = await waitForComponent(
      <SelectMultiValues
        items={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        disabled
        onSearch={onSearch}
        onRemove={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSearch).not.toHaveBeenCalled();

    const search = getByRole(container, 'textbox');
    expect(search).toBeInTheDocument();

    // search query string
    act(() => {
      fireEvent.change(search, { target: { value: 'QUERY' } });
      jest.runAllTimers();
    });

    expect(onSearch).not.toHaveBeenCalled();

    jest.useRealTimers();
  });
});
