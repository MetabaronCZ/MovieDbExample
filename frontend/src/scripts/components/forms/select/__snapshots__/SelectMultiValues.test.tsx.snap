// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/forms/select/SelectMultiValues should not render at all when no options and onSearch not defined 1`] = `<div />`;

exports[`components/forms/select/SelectMultiValues should not render search input when onSearch not defined 1`] = `
.c2 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c2:disabled {
  cursor: not-allowed;
}

.c0 {
  list-style-type: none;
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-size: 12px;
  line-height: 24px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 0 8px;
  background: #98c3d9;
  white-space: nowrap;
}

.c3:hover:not(:disabled) {
  background-color: #0af;
}

.c3:focus {
  outline: 1px dashed #002d56;
}

.c3:disabled {
  cursor: initial;
}

.c4 {
  width: 24px;
  height: 24px;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        type="button"
      >
        Item 1
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        type="button"
      >
        Item 2
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
  </ul>
</div>
`;

exports[`components/forms/select/SelectMultiValues should render properly 1`] = `
.c2 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c2:disabled {
  cursor: not-allowed;
}

.c6 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
}

.c6::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c6:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c6:disabled {
  color: #98c3d9;
}

.c0 {
  list-style-type: none;
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c5 {
  flex: 1;
  padding-left: 4px;
}

.c7 {
  width: 100%;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-size: 12px;
  line-height: 24px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 0 8px;
  background: #98c3d9;
  white-space: nowrap;
}

.c3:hover:not(:disabled) {
  background-color: #0af;
}

.c3:focus {
  outline: 1px dashed #002d56;
}

.c3:disabled {
  cursor: initial;
}

.c4 {
  width: 24px;
  height: 24px;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        title="DESC 1"
        type="button"
      >
        Item 1
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        title="DESC 2"
        type="button"
      >
        Item 2
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
    <li
      class="c1 c5"
    >
      <input
        autocomplete="one-time-code"
        class="c6 c7"
        id="ID"
        placeholder="Hledat"
        type="text"
        value=""
      />
    </li>
  </ul>
</div>
`;

exports[`components/forms/select/SelectMultiValues should render wrapped variant 1`] = `
.c2 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c2:disabled {
  cursor: not-allowed;
}

.c0 {
  list-style-type: none;
  flex: 1;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-size: 12px;
  line-height: 24px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 0 8px;
  background: #98c3d9;
  white-space: nowrap;
}

.c3:hover:not(:disabled) {
  background-color: #0af;
}

.c3:focus {
  outline: 1px dashed #002d56;
}

.c3:disabled {
  cursor: initial;
}

.c4 {
  width: 24px;
  height: 24px;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        type="button"
      >
        Item 1
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
    <li
      class="c1"
    >
      <button
        class="c2 c3"
        type="button"
      >
        Item 2
      </button>
      <button
        class="c2 c3 c4"
        title="Odebrat"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-sm "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </li>
  </ul>
</div>
`;
